### PR TITLE
Backports for 5.2.x (BZ1378875 and BZ1378877)

### PR DESCRIPTION
**Pull request by pruivo**
https://bugzilla.redhat.com/show_bug.cgi?id=1378875
https://bugzilla.redhat.com/show_bug.cgi?id=1378877

### CODE REVIEW

File: core/src/main/java/org/infinispan/transaction/TransactionTable.java [hunk 1]
```diff
@@ -532,40 +524,82 @@ private void shutDownGracefully() {
...
+        if (completedTransactionsInfo != null && !completedTransactionsInfo.isEmpty()) {
+            try {
+                log.tracef("About to cleanup completed transaction. Initial size is %d", completedTransactionsInfo.size());
+                long beginning = System.nanoTime();
+                long minCompleteTimestamp = System.nanoTime()
+                        - TimeUnit.MILLISECONDS.toNanos(configuration.transaction().completedTxTimeout());
+                int removedEntries = 0;
+
+                // this iterator is weekly consistent and will never throw ConcurrentModificationException
+                Iterator<Map.Entry<Address, CompletedTransactionsPerNode>> nodeIterator = completedTransactionsInfo.entrySet()
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/transaction/TransactionTable.java]: You could just iterate `values()`

File: core/src/main/java/org/infinispan/transaction/TransactionTable.java [hunk 2]
```diff
@@ -532,40 +524,82 @@ private void shutDownGracefully() {
...
+                            removedEntries++;
+                        }
+                    }
+                }
             long duration = System.nanoTime() - beginning;
 
             log.tracef("Finished cleaning up completed transactions. %d transactions were removed, total duration was %d millis, " +
                   "current number of completed transactions is %d", removedEntries, TimeUnit.NANOSECONDS.toMillis(duration),
-                  completedTransactions.size());
+                  completedTransactionsInfo.size());
```
Comment (rvansa) [inline core/src/main/java/org/infinispan/transaction/TransactionTable.java]: I think this is wrong, `completedTransactionsInfo.size()` is just number of nodes, not transactions. The same with the trace log on the beginning.

File: core/src/main/java/org/infinispan/transaction/TransactionTable.java [hunk 3]
```diff
@@ -532,40 +524,82 @@ private void shutDownGracefully() {
     * on a remote node. This might cause leaks, e.g. if the transaction is prepared, committed and prepared again.
     * Once marked as completed (because of commit or rollback) any further prepare received on that transaction are discarded.
     */
-   public void markTransactionCompleted(GlobalTransaction globalTx) {
-      completedTransactions.put(globalTx, System.nanoTime());
-   }
+    public void markTransactionCompleted(GlobalTransaction globalTx) {
+        if (completedTransactionsInfo != null) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/transaction/TransactionTable.java]: no, why?

File: core/src/main/java/org/infinispan/transaction/TransactionTable.java [hunk 4]
```diff
@@ -532,40 +524,82 @@ private void shutDownGracefully() {
...
-            long beginning = System.nanoTime();
-            while (iterator.hasNext()) {
-               Map.Entry<GlobalTransaction, Long> e = iterator.next();
-               long ageNanos = System.nanoTime() - e.getValue();
-               if (TimeUnit.NANOSECONDS.toMillis(ageNanos) >= timeout) {
-                  iterator.remove();
-                  removedEntries++;
-               }
-            }
+        if (completedTransactionsInfo != null && !completedTransactionsInfo.isEmpty()) {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/transaction/TransactionTable.java]: can you explain what you mean?

File: core/src/main/java/org/infinispan/transaction/TransactionTable.java (hunk 3 above)
Comment (rvansa) [inline core/src/main/java/org/infinispan/transaction/TransactionTable.java]: after it is started, the `completedTransactionsInfo` can't be null - why the check then?

File: core/src/main/java/org/infinispan/transaction/TransactionTable.java (hunk 4 above)
Comment (rvansa) [inline core/src/main/java/org/infinispan/transaction/TransactionTable.java]:
```
if (foo) {
   indented stuff
}
```
vs.
```
if (!foo) return;
stuff not indented;
```
Makes the code a bit more readable, imo

File: core/src/main/java/org/infinispan/transaction/TransactionTable.java (hunk 3 above)
Comment (pruivo) [inline core/src/main/java/org/infinispan/transaction/TransactionTable.java]: ```completedTransactionsInfo``` is only set for clustered caches. It is ```null``` for local transactional caches.

**Comment by danberindei** [comment]
@pruivo @rvansa I actually changed it to 2 maps soon after: https://github.com/infinispan/infinispan/commit/8728568b6642d3a9e7177c84a4d3ad515d10e36e
https://issues.jboss.org/browse/ISPN-4293

I think that commit should be included in the backport.

**Comment by pruivo** [comment]
@danberindei done!

**Comment by rvansa** [comment]
OK, all my comments are addressed.

**Comment by danberindei** [comment]
Integrated, thanks Pedro!

**Pull request closed without merging**